File: agent_arena/arena/world.py
"""Arena World - main orchestrator for agent interactions."""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional, Callable, Any
import asyncio
import random
import logging

from ..core.types import AgentStatus, ScheduleMode
from ..core.events import Event, EventBus
from ..core.message import Message, MessageType
from ..agents.agent import Agent
from .registry import AgentRegistry
from .channel import Channel

logger = logging.getLogger(__name__)


@dataclass
class ArenaWorld:
    """
    Main orchestrator for the agent simulation.

    Manages agents, channels, message flow, and scheduling.
    """

    name: str = "Agent Arena"

    # Components
    registry: AgentRegistry = field(default_factory=AgentRegistry)
    channels: Dict[str, Channel] = field(default_factory=dict)
    event_bus: EventBus = field(default_factory=EventBus)

    # Scheduling
    mode: ScheduleMode = ScheduleMode.HYBRID
    round_interval: float = 5.0  # Seconds between rounds
    max_speakers_per_round: int = 3

    # State
    running: bool = False
    current_round: int = 0
    start_time: Optional[datetime] = None

    # Default channel
    default_channel: str = "general"

    # Task handles
    _scheduler_task: Optional[asyncio.Task] = field(default=None, repr=False)

    def __post_init__(self):
        """Initialize default channel."""
        if self.default_channel not in self.channels:
            self.channels[self.default_channel] = Channel(
                name=self.default_channel,
                description="General discussion"
            )

    # === Agent Management ===

    async def add_agent(self, agent: Agent, channels: List[str] = None) -> None:
        """Add an agent to the arena."""
        # Register
        self.registry.register(agent)

        # Connect the agent's SDK client
        await agent.connect()

        # Join channels
        channel_names = channels or [self.default_channel]
        for channel_name in channel_names:
            if channel_name in self.channels:
                self.channels[channel_name].add_member(agent.id)

        # Emit join event
        self.event_bus.emit(Event(
            type="agent_joined",
            data={"agent_id": agent.id, "agent_name": agent.name}
        ))

        # Broadcast join message
        join_msg = Message(
            sender_id="system",
            sender_name="System",
            content=f"{agent.name} has joined the chat",
            channel=self.default_channel,
            type=MessageType.JOIN
        )
        await self.broadcast(join_msg)

        logger.info(f"Agent {agent.name} joined the arena")

    async def remove_agent(self, agent_id: str) -> Optional[Agent]:
        """Remove an agent from the arena."""
        agent = self.registry.get(agent_id)
        if not agent:
            return None

        # Broadcast leave message
        leave_msg = Message(
            sender_id="system",
            sender_name="System",
            content=f"{agent.name} has left the chat",
            channel=self.default_channel,
            type=MessageType.LEAVE
        )
        await self.broadcast(leave_msg)

        # Leave all channels
        for channel in self.channels.values():
            channel.remove_member(agent_id)

        # Disconnect agent
        await agent.disconnect()

        # Unregister
        self.registry.unregister(agent_id)

        # Emit leave event
        self.event_bus.emit(Event(
            type="agent_left",
            data={"agent_id": agent_id, "agent_name": agent.name}
        ))

        logger.info(f"Agent {agent.name} left the arena")
        return agent

    # === Channel Management ===

    def create_channel(self, name: str, description: str = "") -> Channel:
        """Create a new channel."""
        channel = Channel(name=name, description=description)
        self.channels[name] = channel
        return channel

    def get_channel(self, name: str) -> Optional[Channel]:
        """Get a channel by name."""
        return self.channels.get(name)

    # === Messaging ===

    async def broadcast(self, message: Message) -> None:
        """Broadcast a message to a channel."""
        channel = self.channels.get(message.channel)
        if not channel:
            logger.warning(f"Channel {message.channel} not found")
            return

        # Add to channel history
        channel.add_message(message)

        # Emit message event
        self.event_bus.emit(Event(
            type="message",
            data={
                "channel": message.channel,
                "message": message.to_dict()
            }
        ))

        logger.debug(f"[{message.channel}] <{message.sender_name}> {message.content}")

    async def inject_message(self, content: str, sender_name: str = "Human",
                             channel: str = None) -> Message:
        """Inject a message from a human or external source."""
        message = Message(
            sender_id="human",
            sender_name=sender_name,
            content=content,
            channel=channel or self.default_channel
        )
        await self.broadcast(message)
        return message

    # === Simulation Control ===

    async def start(self) -> None:
        """Start the simulation."""
        if self.running:
            return

        self.running = True
        self.start_time = datetime.now()

        # Start event bus
        await self.event_bus.start()

        # Start scheduler
        self._scheduler_task = asyncio.create_task(self._run_scheduler())

        self.event_bus.emit(Event(
            type="simulation_started",
            data={"mode": self.mode.value}
        ))

        logger.info(f"Simulation started in {self.mode.value} mode")

    async def stop(self) -> None:
        """Stop the simulation."""
        if not self.running:
            return

        self.running = False

        # Stop scheduler
        if self._scheduler_task:
            self._scheduler_task.cancel()
            try:
                await self._scheduler_task
            except asyncio.CancelledError:
                pass

        # Stop event bus
        await self.event_bus.stop()

        self.event_bus.emit(Event(
            type="simulation_stopped",
            data={"rounds": self.current_round}
        ))

        logger.info("Simulation stopped")

    async def _run_scheduler(self) -> None:
        """Main scheduler loop."""
        while self.running:
            try:
                if self.mode == ScheduleMode.TURN_BASED:
                    await self._run_round()
                    await asyncio.sleep(self.round_interval)

                elif self.mode == ScheduleMode.ASYNC:
                    await self._check_async_responses()
                    await asyncio.sleep(1.0)

                else:  # HYBRID
                    await self._run_hybrid_step()
                    await asyncio.sleep(1.0)

            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Scheduler error: {e}")
                await asyncio.sleep(1.0)

    async def _run_round(self) -> None:
        """Execute a turn-based round."""
        self.current_round += 1

        self.event_bus.emit(Event(
            type="round_started",
            data={"round": self.current_round}
        ))

        # Get agents who want to speak
        speakers = await self._select_speakers()

        # Have each speaker respond
        for agent in speakers[:self.max_speakers_per_round]:
            await self._agent_respond(agent)

        self.event_bus.emit(Event(
            type="round_ended",
            data={"round": self.current_round, "speakers": len(speakers)}
        ))

    async def _check_async_responses(self) -> None:
        """Check if any agents want to respond asynchronously."""
        channel = self.channels[self.default_channel]
        recent = channel.get_recent_messages(1)

        if not recent:
            return

        last_message = recent[-1]

        # Check for direct mentions
        for agent in self.registry.all():
            if agent.status != AgentStatus.IDLE:
                continue

            if agent.name.lower() in [m.lower() for m in last_message.mentions]:
                await self._agent_respond(agent)

    async def _run_hybrid_step(self) -> None:
        """Run one step of hybrid scheduling."""
        channel = self.channels[self.default_channel]
        recent = channel.get_recent_messages(1)

        # Check for urgent responses (mentions)
        if recent:
            last_message = recent[-1]
            for agent in self.registry.all():
                if agent.status != AgentStatus.IDLE:
                    continue
                if agent.name.lower() in [m.lower() for m in last_message.mentions]:
                    await self._agent_respond(agent)
                    return

        # Periodic round check
        if not hasattr(self, '_last_round_time'):
            self._last_round_time = datetime.now()

        elapsed = (datetime.now() - self._last_round_time).total_seconds()
        if elapsed >= self.round_interval:
            await self._run_round()
            self._last_round_time = datetime.now()

    async def _select_speakers(self) -> List[Agent]:
        """Select agents who will speak this round."""
        channel = self.channels[self.default_channel]
        recent = channel.get_recent_messages(1)

        if not recent:
            # No messages yet - pick a random agent to start
            agents = self.registry.all()
            if agents:
                return [random.choice(agents)]
            return []

        last_message = recent[-1]
        candidates = []

        for agent in self.registry.all():
            if agent.status != AgentStatus.IDLE:
                continue
            if agent.id == last_message.sender_id:
                continue  # Skip the last speaker

            probability = agent.should_respond(last_message, self.registry.names())
            if random.random() < probability:
                candidates.append((probability, agent))

        # Sort by probability and return top speakers
        candidates.sort(key=lambda x: x[0], reverse=True)
        return [agent for _, agent in candidates]

    async def _agent_respond(self, agent: Agent) -> Optional[Message]:
        """Have an agent generate and send a response."""
        channel = self.channels[self.default_channel]

        # Build context from recent messages
        context = f"""Current conversation in #{channel.name}:

{channel.get_context_string(20)}

Participants: {', '.join(self.registry.names())}

Now respond naturally as {agent.name}. Keep it brief (1-2 sentences).
IMPORTANT: Just write your response directly. Do NOT include your name, timestamps, or angle brackets."""

        # Emit thinking event
        self.event_bus.emit(Event(
            type="agent_thinking",
            data={"agent_id": agent.id, "agent_name": agent.name, "thinking": True}
        ))

        # Get response from agent
        response_text = await agent.respond(context)

        # Emit thinking done event
        self.event_bus.emit(Event(
            type="agent_thinking",
            data={"agent_id": agent.id, "agent_name": agent.name, "thinking": False}
        ))

        if not response_text:
            return None

        # Create and broadcast message
        message = Message(
            sender_id=agent.id,
            sender_name=agent.name,
            content=response_text,
            channel=channel.name
        )
        await self.broadcast(message)

        return message

    # === Status ===

    def get_status(self) -> dict:
        """Get current simulation status."""
        return {
            "name": self.name,
            "running": self.running,
            "mode": self.mode.value,
            "current_round": self.current_round,
            "start_time": self.start_time.isoformat() if self.start_time else None,
            "agents": {
                "count": self.registry.count(),
                "names": self.registry.names()
            },
            "channels": {
                name: ch.to_dict() for name, ch in self.channels.items()
            }
        }

File: src/arena/world.ts
/**
 * Arena World - main orchestrator for agent interactions.
 * Uses setInterval instead of asyncio.Task for cleaner scheduling.
 * Now with SQLite persistence for sessions and event logging.
 */

import { v4 as uuidv4 } from 'uuid';
import { ScheduleMode, AgentStatus, MessageType, WorldStatus, Attachment } from '../core/types.js';
import { EventBus } from '../core/events.js';
import { Message } from '../core/message.js';
import { Agent } from '../agents/agent.js';
import { AgentRegistry } from './registry.js';
import { Channel } from './channel.js';
import * as db from '../core/database.js';

export class ArenaWorld {
  name: string;
  registry: AgentRegistry;
  channels: Map<string, Channel>;
  eventBus: EventBus;
  mode: ScheduleMode;
  roundInterval: number; // seconds between rounds
  maxSpeakersPerRound: number;
  running: boolean;
  currentRound: number;
  startTime: Date | null;
  defaultChannel: string;
  sessionId: string | null = null;
  totalMessagesInSession: number = 0;
  maxTurns: number = 20; // Auto-stop after this many rounds

  private schedulerInterval: NodeJS.Timeout | null = null;
  private lastRoundTime: Date = new Date();

  constructor(name: string = 'Agent Arena') {
    this.name = name;
    this.registry = new AgentRegistry();
    this.channels = new Map();
    this.eventBus = new EventBus();
    this.mode = ScheduleMode.HYBRID;
    this.roundInterval = 5.0;
    this.maxSpeakersPerRound = 3;
    this.running = false;
    this.currentRound = 0;
    this.startTime = null;
    this.defaultChannel = 'general';

    // Try to load existing default channel from database, otherwise create new
    const existingRoom = db.getRoom(this.defaultChannel);
    if (existingRoom) {
      const channel = Channel.fromDatabase(existingRoom.id);
      if (channel) {
        this.channels.set(this.defaultChannel, channel);
        console.log(`Loaded existing room '${this.defaultChannel}' from database`);
      } else {
        this.channels.set(this.defaultChannel, new Channel(
          this.defaultChannel,
          'General discussion'
        ));
      }
    } else {
      this.channels.set(this.defaultChannel, new Channel(
        this.defaultChannel,
        'General discussion'
      ));
    }
  }

  // === Agent Management ===

  async addAgent(agent: Agent, channelNames?: string[]): Promise<void> {
    // Register
    this.registry.register(agent);

    // Connect the agent's SDK client
    await agent.connect();

    // Join channels
    const channels = channelNames || [this.defaultChannel];
    for (const channelName of channels) {
      const channel = this.channels.get(channelName);
      if (channel) {
        channel.addMember(agent.id);
      }
    }

    // Emit join event
    this.eventBus.emitEvent('agent_joined', {
      agent_id: agent.id,
      agent_name: agent.name
    });

    // Broadcast join message
    const joinMsg = new Message(
      'system',
      'System',
      `${agent.name} has joined the chat`,
      this.defaultChannel,
      MessageType.JOIN
    );
    await this.broadcast(joinMsg);

    console.log(`Agent ${agent.name} joined the arena`);
  }

  async removeAgent(agentId: string): Promise<Agent | undefined> {
    const agent = this.registry.get(agentId);
    if (!agent) {
      return undefined;
    }

    // Broadcast leave message
    const leaveMsg = new Message(
      'system',
      'System',
      `${agent.name} has left the chat`,
      this.defaultChannel,
      MessageType.LEAVE
    );
    await this.broadcast(leaveMsg);

    // Leave all channels
    for (const channel of this.channels.values()) {
      channel.removeMember(agentId);
    }

    // Disconnect agent
    await agent.disconnect();

    // Unregister
    this.registry.unregister(agentId);

    // Emit leave event
    this.eventBus.emitEvent('agent_left', {
      agent_id: agentId,
      agent_name: agent.name
    });

    console.log(`Agent ${agent.name} left the arena`);
    return agent;
  }

  // === Channel Management ===

  createChannel(name: string, description: string = ''): Channel {
    const channel = new Channel(name, description);
    this.channels.set(name, channel);
    return channel;
  }

  getChannel(name: string): Channel | undefined {
    return this.channels.get(name);
  }

  // === Messaging ===

  async broadcast(message: Message): Promise<void> {
    const channel = this.channels.get(message.channel);
    if (!channel) {
      console.warn(`Channel ${message.channel} not found`);
      return;
    }

    // Add to channel history (this persists to database)
    channel.addMessage(message);

    // Increment session message count
    if (this.sessionId) {
      this.totalMessagesInSession++;
    }

    // Log event to database
    db.logEvent('message', {
      message_id: message.id,
      channel: message.channel,
      sender_id: message.sender_id,
      sender_name: message.sender_name,
      content_length: message.content.length,
      mentions: message.mentions
    }, this.sessionId || undefined);

    // Emit message event
    this.eventBus.emitEvent('message', {
      channel: message.channel,
      message: message.toDict()
    });

    console.log(`[${message.channel}] <${message.sender_name}> ${message.content}`);
  }

  async injectMessage(
    content: string,
    senderName: string = 'Human',
    channelName?: string,
    attachments?: Attachment[]
  ): Promise<Message> {
    const message = new Message(
      'human',
      senderName,
      content,
      channelName || this.defaultChannel,
      MessageType.CHAT,
      undefined,
      undefined,
      undefined,
      undefined,
      attachments
    );
    await this.broadcast(message);
    return message;
  }

  // === Simulation Control ===

  start(): void {
    if (this.running) {
      return;
    }

    this.running = true;
    this.startTime = new Date();
    this.lastRoundTime = new Date();
    this.totalMessagesInSession = 0;

    // Create session in database
    this.sessionId = uuidv4();
    db.createSession(this.sessionId, this.name, this.mode);
    db.logEvent('simulation_started', { mode: this.mode, session_id: this.sessionId }, this.sessionId);

    // Start scheduler - runs every second
    this.schedulerInterval = setInterval(() => {
      this.tick().catch(err => {
        console.error('Scheduler error:', err);
      });
    }, 1000);

    this.eventBus.emitEvent('simulation_started', {
      mode: this.mode,
      session_id: this.sessionId
    });

    console.log(`Simulation started in ${this.mode} mode (session: ${this.sessionId})`);
  }

  stop(): void {
    if (!this.running) {
      return;
    }

    this.running = false;

    // Stop scheduler
    if (this.schedulerInterval) {
      clearInterval(this.schedulerInterval);
      this.schedulerInterval = null;
    }

    // End session in database
    if (this.sessionId) {
      db.endSession(this.sessionId, this.currentRound, this.totalMessagesInSession);
      db.logEvent('simulation_stopped', {
        rounds: this.currentRound,
        messages: this.totalMessagesInSession
      }, this.sessionId);
    }

    this.eventBus.emitEvent('simulation_stopped', {
      rounds: this.currentRound,
      session_id: this.sessionId
    });

    console.log(`Simulation stopped (${this.currentRound} rounds, ${this.totalMessagesInSession} messages)`);
    this.sessionId = null;
  }

  private async tick(): Promise<void> {
    if (!this.running) return;

    // Check if we've hit max turns
    if (this.maxTurns > 0 && this.currentRound >= this.maxTurns) {
      console.log(`Max turns (${this.maxTurns}) reached, stopping simulation`);
      this.stop();
      return;
    }

    switch (this.mode) {
      case ScheduleMode.TURN_BASED:
        await this.tickTurnBased();
        break;
      case ScheduleMode.ASYNC:
        await this.checkAsyncResponses();
        break;
      case ScheduleMode.HYBRID:
        await this.runHybridStep();
        break;
    }
  }

  private async tickTurnBased(): Promise<void> {
    const elapsed = (Date.now() - this.lastRoundTime.getTime()) / 1000;
    if (elapsed >= this.roundInterval) {
      await this.runRound();
      this.lastRoundTime = new Date();
    }
  }

  private async runRound(): Promise<void> {
    this.currentRound += 1;

    this.eventBus.emitEvent('round_started', {
      round: this.currentRound
    });

    // Get agents who want to speak
    const speakers = await this.selectSpeakers();

    // Have each speaker respond
    for (const agent of speakers.slice(0, this.maxSpeakersPerRound)) {
      await this.agentRespond(agent);
    }

    this.eventBus.emitEvent('round_ended', {
      round: this.currentRound,
      speakers: speakers.length
    });
  }

  private async checkAsyncResponses(): Promise<void> {
    const channel = this.channels.get(this.defaultChannel);
    if (!channel) return;

    const recent = channel.getRecentMessages(1);
    if (recent.length === 0) return;

    const lastMessage = recent[recent.length - 1];

    // Check for direct mentions
    for (const agent of this.registry.all()) {
      // Allow OFFLINE (not yet connected) or IDLE agents to respond
      if (agent.status !== AgentStatus.IDLE && agent.status !== AgentStatus.OFFLINE) {
        continue;
      }

      const mentionsLower = lastMessage.mentions.map(m => m.toLowerCase());
      if (mentionsLower.includes(agent.name.toLowerCase())) {
        await this.agentRespond(agent);
      }
    }
  }

  private async runHybridStep(): Promise<void> {
    const channel = this.channels.get(this.defaultChannel);
    if (!channel) return;

    const recent = channel.getRecentMessages(1);

    // Check for urgent responses (mentions)
    if (recent.length > 0) {
      const lastMessage = recent[recent.length - 1];
      for (const agent of this.registry.all()) {
        // Allow OFFLINE (not yet connected) or IDLE agents to respond
        if (agent.status !== AgentStatus.IDLE && agent.status !== AgentStatus.OFFLINE) {
          continue;
        }
        const mentionsLower = lastMessage.mentions.map(m => m.toLowerCase());
        if (mentionsLower.includes(agent.name.toLowerCase())) {
          await this.agentRespond(agent);
          return;
        }
      }
    }

    // Periodic round check
    const elapsed = (Date.now() - this.lastRoundTime.getTime()) / 1000;
    if (elapsed >= this.roundInterval) {
      await this.runRound();
      this.lastRoundTime = new Date();
    }
  }

  private async selectSpeakers(): Promise<Agent[]> {
    const channel = this.channels.get(this.defaultChannel);
    if (!channel) return [];

    const recent = channel.getRecentMessages(1);

    if (recent.length === 0) {
      // No messages yet - pick a random agent to start
      const agents = this.registry.all();
      if (agents.length > 0) {
        const randomIndex = Math.floor(Math.random() * agents.length);
        return [agents[randomIndex]];
      }
      return [];
    }

    const lastMessage = recent[recent.length - 1];
    const candidates: Array<[number, Agent]> = [];

    for (const agent of this.registry.all()) {
      // Allow OFFLINE (not yet connected) or IDLE agents to respond
      if (agent.status !== AgentStatus.IDLE && agent.status !== AgentStatus.OFFLINE) {
        continue;
      }
      if (agent.id === lastMessage.sender_id) {
        continue; // Skip the last speaker
      }

      const probability = agent.shouldRespond(lastMessage, this.registry.names());
      if (Math.random() < probability) {
        candidates.push([probability, agent]);
      }
    }

    // Sort by probability and return top speakers
    candidates.sort((a, b) => b[0] - a[0]);
    return candidates.map(([, agent]) => agent);
  }

  async agentRespond(agent: Agent): Promise<Message | null> {
    const channel = this.channels.get(this.defaultChannel);
    if (!channel) return null;

    // Build context from recent messages
    const context = `Current conversation in #${channel.name}:

${channel.getContextString(20)}

Participants: ${this.registry.names().join(', ')}

Now respond naturally as ${agent.name}. Keep it brief (1-2 sentences).
IMPORTANT: Just write your response directly. Do NOT include your name, timestamps, or angle brackets.`;

    // Emit thinking event
    this.eventBus.emitEvent('agent_thinking', {
      agent_id: agent.id,
      agent_name: agent.name,
      thinking: true
    });

    // Log agent thinking start
    db.logEvent('agent_thinking_start', {
      agent_id: agent.id,
      agent_name: agent.name
    }, this.sessionId || undefined);

    // Get response from agent
    const responseText = await agent.respond(context);

    // Log agent thinking end
    db.logEvent('agent_thinking_end', {
      agent_id: agent.id,
      agent_name: agent.name,
      responded: !!responseText
    }, this.sessionId || undefined);

    // Emit thinking done event
    this.eventBus.emitEvent('agent_thinking', {
      agent_id: agent.id,
      agent_name: agent.name,
      thinking: false
    });

    if (!responseText) {
      return null;
    }

    // Create and broadcast message
    const message = new Message(
      agent.id,
      agent.name,
      responseText,
      channel.name
    );
    await this.broadcast(message);

    return message;
  }

  // === Status ===

  getStatus(): WorldStatus {
    const channelsData: Record<string, ReturnType<Channel['toDict']>> = {};
    for (const [name, channel] of this.channels) {
      channelsData[name] = channel.toDict();
    }

    return {
      name: this.name,
      running: this.running,
      mode: this.mode,
      current_round: this.currentRound,
      max_turns: this.maxTurns,
      start_time: this.startTime?.toISOString() || null,
      agents: {
        count: this.registry.count(),
        names: this.registry.names()
      },
      channels: channelsData
    };
  }
}
